android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/services/LocationService.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021-2024 Ona Systems, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.rulesengine.services
+
+import android.content.Context
+import android.location.Location
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationServices
+import dagger.hilt.android.qualifiers.ApplicationContext
+import java.time.Instant
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.Resource
+import org.smartregister.fhircore.engine.util.SharedPreferenceKey
+import org.smartregister.fhircore.engine.util.SharedPreferencesHelper
+import org.smartregister.fhircore.engine.util.location.LocationUtils.getAccurateLocation
+import org.smartregister.fhircore.engine.util.location.LocationUtils.getApproximateLocation
+import org.smartregister.fhircore.engine.util.location.PermissionUtils.hasCoarseLocationPermissions
+import org.smartregister.fhircore.engine.util.location.PermissionUtils.hasFineLocationPermissions
+
+class LocationService(
+  @ApplicationContext val context: Context,
+  val sharedPreferences: SharedPreferencesHelper,
+) {
+  lateinit var fusedLocationProviderClient: FusedLocationProviderClient
+
+  fun calculateDistanceByProvidedLocations(
+    destination: Location,
+    currentLocation: Location,
+  ): String {
+    val distanceInMeters = currentLocation.distanceTo(destination)
+    return formatDistance(distanceInMeters)
+  }
+
+  fun calculateDistanceByGpsLocation(location: Resource): String {
+    val currentLocation = generateLocation(location)
+
+    CoroutineScope(Dispatchers.IO).launch {
+      fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
+
+      val retrievedLocation =
+        if (hasFineLocationPermissions(context)) {
+          getAccurateLocation(fusedLocationProviderClient, Dispatchers.IO)
+        } else if (hasCoarseLocationPermissions(context)) {
+          getApproximateLocation(fusedLocationProviderClient, Dispatchers.IO)
+        } else {
+          null
+        }
+
+      retrievedLocation?.let {
+        writeLocation(LocationCoordinates(it.latitude, it.longitude, it.altitude, Instant.now()))
+      }
+    }
+
+    val locationTest =
```

Comment:
What does `locationTest` mean?

Reply:
Rename this to something more relevant. Not sure if my suggested name will suffice.